### DISCUSSION TITLE
Add support for FreeRTOS

### DISCUSSION THREAD

**Discussion by gtxzsxxk** [gtxzsxxk/duo-buildroot#10] [Ideas]
Recently I have added a package which will download and compile the modified version of `FreeRTOS` released by the official duo-buildroot-sdk. I fix some gcc compatible problems when compiling with buildroot toolchains. Now when we say Y to this package, we could directly treat the `output/build/milkv-duo-smallcore-freertos-hash` as our project directory. We can add our code and the buildroot will automatically compile and pack this into the `fip.bin` and `sdcard.img`, loading the `FreeRTOS` into the small-core.

Do you have any ideas on that?

**Comment by shodan8192** [comment]
Haven't looked it yet, if something come to mind, I'll let you know.

**Reply by gtxzsxxk** [reply]
OK. Thanks!

**Reply by shodan8192** [reply]
Approach with `output/build/milkv-duo-smallcore-freertos-hash/cvitek/task/comm/src/riscv64/` as project directory has some downsides :
- `hash` name part is variable, will change between commits
- it's burried deep inside `output` folder
- we have to make whole buildroot at least once
- it is volatile, `make clean all` will vanish our code from there

Maybe FreeRTOS package can be patched here :
https://github.com/gtxzsxxk/milkv-duo-smallcore-freertos/blob/fb1fb0f6061a8a3397adaa90b5c1546527156313/cvitek/task/CMakeLists.txt#L23
to point some other directory, eg inside `board/milkv/duo` ?

**Comment by gtxzsxxk** [comment]
Thank you for your review~ OK, let me explain them.

> hash name part is variable, will change between commits

Sure. So I have to try keeping the github repo unchanged.

> it's burried deep inside output folder
> it is volatile, make clean all will vanish our code from there

That's right. I am looking for better solutions as well.

> we have to make whole buildroot at least once

We could use `make milkv-duo-smallcore-freertos-rebuild` to simply rebuilt our freertos. And if we do not have much changes, when `make` the buildroot, it will just simply do some pack things very quickly.

**Comment by gtxzsxxk** [comment]
And it seems that there are no more better way to write code for the freertos. If you are using the official sdk, you have to make the whole project too.

**Reply by shodan8192** [reply]
You are right of all about this. So it is not possible to change directory path that's holds user code to something more convenient ?

**Reply by gtxzsxxk** [reply]
Maybe possible. Maybe we can try add a prompt in the buildroot, asking the user to provide their code. When we make the whole buildroot or rebuild the freertos package, the mk script will copy the user's code and then compile. This is just my imagination. It seems possible but I don't know if it is practical. Maybe I can give it a try after I survive my midterm exam.

**Reply by shodan8192** [reply]
OK, good luck with the exam then !

**Reply by gtxzsxxk** [reply]
haha, thanks for your help and ideas too!
